### PR TITLE
Fix recursive bugs

### LINKED ISSUE #30: Make use of `globby.generateGlobTasks`
// @schnittstabil

### LINKED ISSUE #73: Cannot Copy Documents Folder on Windows
https://github.com/jprichardson/node-fs-extra/issues/726
I am not able to copy JUNCTION from windows

image
image

### LINKED ISSUE #91: Strange behavior on latest `recursive` mode
I installed latest `cpy` with `yarn add cpy@sindresorhus/cpy`.

```js
// copy.js

// @ts-check

import cpy from 'cpy'

const args = process.argv.slice(2)

cpy(args, args.pop())
```

```sh
node copy 'node_modules/shiki/{dist/*.wasm,languages,themes}/**' public/shiki
```

image

```sh
node copy 'node_modules/shiki/{dist/*.wasm,languages,themes}/**' public/shiki/_ # note the `/_` here
```

image

I also tried the following one and it failed.

```sh
node copy 'node_modules/shiki/{dist/*.wasm,languages/**,themes/**}' public/shiki
```

But the following one also worked

```sh
node copy 'node_modules/shiki/{dist/*.wasm,languages/**,themes/**}' public/shiki/_
```

### PR DESCRIPTION
**Pull request by Idered**
This PR fixes 5 bugs:
- fixes #91
- fixes #73
- absolute directory source paths not working
- absolute file source paths not working
- negative patterns not working

This issue can be closed too:
- closes #30 

This PR completes my work on recursive feature

I'm also attaching a small migration guide bellow.

Can I get `hacktoberfest-accepted` label on this?

---

# Breaking changes & examples

## Recursive by default

- `parents` option was removed 
- to get flat list of files use `flat: true` option

Example directory structure:
```
- .github/workflows/main.yml
- .github/funding.yml
```

Command: `cpy('.github/**', 'dest')`

Old output:
```
- dest/funding.yml
- dest/main.yml
```
New output: 
```
- dest/workflows/main.yml
- dest/main.yml
```

## Recreate old `parents: true`

Example directory structure:
```
- .github/workflows/main.yml
- .github/funding.yml
```

Old:
`cpy('.github/**', 'dest', {parents: true})`

New:
`cpy('.github', 'dest')`

Output:
```
- dest/.github/workglows/main.yml
- dest/.github/funding.yml
```

## Recreate old `parents: false`
Example directory structure:
```
- .github/workflows/main.yml
- .github/funding.yml
```

Old:
`cpy('.github/**', 'dest', {parents: false})`

New:
`cpy('.github', 'dest', {flat: true})`

Output:
```
- dest/main.yml
- dest/funding.yml
```

## Copy all package.json and preserve folder structure

Old: 
`cpy('node_modules/**/package.json', 'dest', {parents: true})`

Old output:

New: 
`cpy('node_modules/**/package.json', 'dest')`

New output:

> Ignore colors, those are git changes

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for working on this. Would you be able to add some tests for the bugs being fixed?

**Comment by sindresorhus** [comment]
> Can I get hacktoberfest-accepted label on this?

✅

File: index.js [hunk 1]
```diff
@@ -55,6 +56,24 @@ class Entry {
...
+@returns {string[]}
+*/
+const expandPatternsWithBraceExpansion = patterns => {
+	let collection = [];
+	for (const pattern of patterns) {
+		collection = [...collection, ...micromatch.braces(pattern, {
+			expand: true,
+			nodupes: true,
+		})];
+	}
```
Comment (sindresorhus) [inline index.js:72]: This would be a good place to use `Array#flatMap`.

File: index.js [hunk 2]
```diff
@@ -82,6 +101,14 @@ const preprocessDestinationPath = ({entry, destination, options}) => {
 		return path.join(destination, entry.name);
 	}
 
+	if (entry.pattern.isDirectory && path.relative(options.cwd, entry.path).startsWith('..')) {
```
Comment (sindresorhus) [inline index.js:104]: This check will not work correctly if `options.cwd` and `entry.path` are on different partitions on Windows, see: https://github.com/sindresorhus/import-local/pull/12

File: index.js [hunk 3]
```diff
@@ -82,6 +101,14 @@ const preprocessDestinationPath = ({entry, destination, options}) => {
...
+		return path.join(options.cwd, destination, path.basename(entry.pattern.originalPath), path.relative(entry.pattern.originalPath, entry.path));
+	}
+
+	if (!entry.pattern.isDirectory && entry.path === entry.relativePath) {
```
Comment (sindresorhus) [inline index.js:108]: It would be good to have a short code comment on both of these `if`'s on what kind of case they handle.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Bump :)

**Comment by iiLearner** [comment]
@Idered would you be able to have a follow up to this as its blocking https://github.com/sindresorhus/cpy/issues/98 :)

File: index.js [hunk 4]
```diff
@@ -55,6 +56,24 @@ class Entry {
 	}
 }
 
+/**
+ * Expand patterns like:
+ * "node_modules/{globby,micromatch}" into ["node_modules/globby", "node_modules/micromatch"]
+@param {string[]} patterns
+@returns {string[]}
+*/
```
Comment (sindresorhus) [inline index.js:64]:
```suggestion
/**
Expand patterns like `'node_modules/{globby,micromatch}'` into `['node_modules/globby', 'node_modules/micromatch']`.

@param {string[]} patterns
@returns {string[]}
*/
```

**Comment by thetutlage** [comment]
This PR introduces a regression. Following is a snippet from the PR original content.

## Recreate old parents: true
Example directory structure:

- `.github/workflows/main.yml`
- `.github/funding.yml`

Old:
```
cpy('.github/**', 'dest', {parents: true})
```

New:
```
cpy('.github', 'dest')
```

Output:

```
- dest/.github/workglows/main.yml
- dest/.github/funding.yml
```

----

How should I have the same output as previous `parents: true` while defining the file extensions as well.

Old:

```ts
cpy('.github/**/*.yml', 'dest', {parents: true})
```

Old Output:

```
- dest/.github/workglows/main.yml
- dest/.github/funding.yml
```

New:

```ts
cpy('.github/**/*.yaml', 'dest')
```

New Output:

```
- dest/workglows/main.yml
- dest/funding.yml
```

**Pull request merged**